xc/device.ml

```diff
@@ -212,20 +212,27 @@ module Generic = struct
         )
       )
 
+  let error_watch ~xs (x: device) = Watch.value_to_appear (error_path_of_device ~xs x)
+  let frontend_closed ~xs (x: device) = Watch.map (fun () -> "") (Watch.value_to_become (frontend_rw_path_of_device ~xs x ^ "/state") (Xenbus_utils.string_of Xenbus_utils.Closed))
+  let backend_closed ~xs (x: device) = Watch.value_to_become (backend_path_of_device ~xs x ^ "/state") (Xenbus_utils.string_of Xenbus_utils.Closed)
+
   let unplug_watch ~xs (x: device) =
     let path = Hotplug.path_written_by_hotplug_scripts x in
     let qdisk = Astring.String.is_infix ~affix:"backend/qdisk/" path in
     if not qdisk then begin
       Watch.key_to_disappear path
     end else begin
-      debug "unplug_watch: not waiting for qdisk %s; returning dummy watch" path;
-      Watch.{ evaluate = fun xs -> try xs.Xs.rm path with _ -> debug "dummy unplug_watch for qdisk: %s already removed" path }
+      (* We need to delete this key ourselves, qemu-dp wouldn't delete it *)
+      debug "unplug_watch: not waiting for qdisk %s; waiting for frontend closed" path;
+      (* Wait until either we've already deleted the key, or the frontend has reached the closed
+       * state *)
+      Watch.any_of [
```

Comment:
I have trouble parsing this code. Could the formatting be improved to highlight the structure?

Reply:
I've added another variable to make it clearer what goes into `Watch.any_of` (it takes 2 other watches, now each one has an explicit name).